IdentityCore/src/util/mac/MSIDAppExtensionUtil.m

```diff
@@ -0,0 +1,95 @@
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#import "MSIDAppExtensionUtil.h"
+#import "MSIDMainThreadUtil.h"
+
+@implementation MSIDAppExtensionUtil
+
++ (BOOL)isExecutingInAppExtension
+{
+    NSString *mainBundlePath = [[NSBundle mainBundle] bundlePath];
+    
+    if (mainBundlePath.length == 0)
+    {
+        MSID_LOG_ERROR(nil, @"Expected `[[NSBundle mainBundle] bundlePath]` to be non-nil. Defaulting to non-application-extension safe API.");
+        
+        return NO;
+    }
+    
+    return [mainBundlePath hasSuffix:@"appex"];
+}
+
+#pragma mark - NSWorkspace
+
++ (NSWorkspace *)sharedApplication
+{
+    if ([self isExecutingInAppExtension])
+    {
+        // The caller should do this check but we will double check to fail safely
+        return nil;
+    }
+    
+    return [NSWorkspace performSelector:NSSelectorFromString(@"sharedWorkspace")];
+}
+
++ (void)sharedApplicationOpenURL:(NSURL*)url
+{
+    if ([self isExecutingInAppExtension])
+    {
+        // The caller should do this check but we will double check to fail safely
+        return;
+    }
+    
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Warc-performSelector-leaks"
+    
+    [MSIDMainThreadUtil executeOnMainThreadIfNeeded:^{
```

Comment:
The API sounds like some UI related event, could you make it run on main thread when calling instead of doing it internally?

Reply:
Copied the existing code from util/ios/MSIDAppExtensionUtil.m. This is meant to be a utility class API used by us as a safe replacement for 'openURL:'. 
What would the benefit be if our calling code calls this API within main thread block as opposed to this API doing it internally? I think not worrying about calling this API in main thread every time it is used is beneficial.

Reply:
No difference from functionality wise, and normally is the caller to decide which thread it could run. Just a nit, and up to your logic